optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -184,24 +169,34 @@ def log_pdf(self, samples_dict: Dict[str, np.ndarray]) -> np.ndarray:
                 assert mus is not None
                 assert sigmas is not None
 
-                cdf_func = _ParzenEstimator._normal_cdf
-                p_accept = cdf_func(high, mus, sigmas) - cdf_func(low, mus, sigmas)
                 if q is None:
-                    distance = samples[:, None] - mus
-                    mahalanobis = distance / np.maximum(sigmas, EPS)
-                    z = np.sqrt(2 * np.pi) * sigmas
-                    coefficient = 1 / z / p_accept
-                    log_pdf = -0.5 * mahalanobis**2 + np.log(coefficient)
+                    log_pdf = truncnorm.logpdf(
+                        samples[:, None],
+                        (low - mus) / sigmas,
+                        (high - mus) / sigmas,
+                        loc=mus,
+                        scale=sigmas,
+                    )
                 else:
+                    # TODO(nzw0301): Simplify the logic by following
+                    # https://github.com/optuna/optuna/pull/3985#issuecomment-1253637444
+                    # when we can assume scipy 1.9.2 is commonly used, which includes
+                    # accurate logcdf https://github.com/scipy/scipy/pull/17064.
+                    cdf_func = _ParzenEstimator._trunc_normal_cdf
+                    p_accept = cdf_func(high, mus, sigmas, low, high) - cdf_func(
+                        low, mus, sigmas, low, high
+                    )
+
                     upper_bound = np.minimum(samples + q / 2.0, high)
                     lower_bound = np.maximum(samples - q / 2.0, low)
-                    cdf = cdf_func(upper_bound[:, None], mus[None], sigmas[None]) - cdf_func(
-                        lower_bound[:, None], mus[None], sigmas[None]
-                    )
+                    cdf = cdf_func(
+                        upper_bound[:, None], mus[None], sigmas[None], low, high
+                    ) - cdf_func(lower_bound[:, None], mus[None], sigmas[None], low, high)
                     log_pdf = np.log(cdf + EPS) - np.log(p_accept + EPS)
```

Comment:
I noticed that it can be rewritten as follows. With this change, the speed is comparable to using SciPy in my environment.

```python
log_gauss_mass = truncnorm._log_gauss_mass((lower_bound[:, None] - mus) / sigmas, (upper_bound[:, None] - mus) / sigmas)
log_p_accept = truncnorm._log_gauss_mass((low - mus) / sigmas, (high - mus) / sigmas)
log_pdf = log_gauss_mass - log_p_accept
```

Reply:
This is blasting fast! The speed was comparable to SciPy version in my environment too.